### PR TITLE
feat(web-analytics): Allow customers to compare against a different period

### PR DESCRIPTION
**Pull request by rafaeelaudibert**
## Changes

Right now, we'd always compare our data against the last period, i.e. if we're looking at the data from the last 14 days, then you'd always compare it against the previous 14 days.

This is not always ideal, as someone might want to compare against numbers from the same time last year.

We've added the comparator component in a previous commit (hidden behind a FF) and we're now actually using that data in the backend.

image

## Does this work well for both Cloud and self-hosted?

Yep

## How did you test this code?

- [x] Manual UI tests 
- [x] Backend tests

### CODE REVIEW

File: frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx [hunk 1]
```diff
@@ -882,6 +882,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                                                   kind: NodeKind.WebExternalClicksTableQuery,
                                                   properties: webAnalyticsFilters,
                                                   dateRange,
+                                                  compareFilter, // NOTE: Backend is not using this yet
```
Comment (rafaeelaudibert) [inline frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:885]: @robbie-c Thoughts on whether this is actually helpful? Does it make sense to compare outbound link clicks? The backend doesn't compare it at all, it's always just current data

It's not hard to implement, but I'm trying to get a feeling here on how this could be helpful. I think it is, but I wanna hear from you.

Comment (robbie-c) [reply frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:885]: Yeah I think it is, but fine if we don't include it in this PR

File: frontend/__snapshots__/scenes-other-onboarding--onboarding-billing--light.png
Comment (rafaeelaudibert) [inline frontend/__snapshots__/scenes-other-onboarding--onboarding-billing--light.png:1]: Come on buddy, why do you keep disappearing 😮‍💨

File: posthog/hogql_queries/web_analytics/external_clicks.py [hunk 2]
```diff
@@ -43,36 +43,38 @@ def to_query(self) -> ast.SelectQuery:
...
         cutToFirstSignificantSubdomain(properties.`$external_click_url`) != cutToFirstSignificantSubdomain(properties.`$host`),
-        {all_properties}
+        {inside_periods},
+        {all_properties},
     )
     GROUP BY events.`$session_id`, url
 )
 GROUP BY "context.columns.url"
 ORDER BY "context.columns.visitors" DESC,
+"context.columns.clicks" DESC,
```
Comment (robbie-c) [inline posthog/hogql_queries/web_analytics/external_clicks.py:68]: nice drive-by, makes sense

File: posthog/hogql_queries/web_analytics/external_clicks.py [hunk 3]
```diff
@@ -43,36 +43,38 @@ def to_query(self) -> ast.SelectQuery:
...
-    sum(filtered_click_count) AS "context.columns.clicks"
+    tuple(uniqIf(filtered_person_id, {current_period}), uniqIf(filtered_person_id, {previous_period})) AS "context.columns.visitors",
+    tuple(sumIf(filtered_click_count, {current_period}), sumIf(filtered_click_count, {previous_period})) AS "context.columns.clicks"
 FROM (
     SELECT
         any(person_id) AS filtered_person_id,
         count() AS filtered_click_count,
-        {url_expr} AS url
+        {url_expr} AS url,
+        MIN(timestamp) AS start_timestamp
```
Comment (robbie-c) [inline posthog/hogql_queries/web_analytics/external_clicks.py:53]: This will be the timestamp of the first click rather than the start of the session. If you want the start of the session you can do `any(session.$start_timestamp) as start_timestamp`

File: posthog/hogql_queries/utils/query_previous_period_date_range.py [hunk 4]
```diff
@@ -15,13 +15,13 @@ class QueryPreviousPeriodDateRange(QueryDateRange):
     """Translation of the raw `date_from` and `date_to` filter values to datetimes."""
 
     _team: Team
-    _date_range: Optional[InsightDateRange]
+    _date_range: Optional[InsightDateRange | DateRange]
```
[the remark this answers is not in the document]
Comment (robbie-c) [reply posthog/hogql_queries/utils/query_previous_period_date_range.py:18]: A: They look the same except for defaults
```python
class InsightDateRange(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )
    date_from: Optional[str] = "-7d"
    date_to: Optional[str] = None
    explicitDate: Optional[bool] = Field(
        default=False,
        description=(
            "Whether the date_from and date_to should be used verbatim. Disables rounding to the start and end of"
            " period."
        ),
    )

class DateRange(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    explicitDate: Optional[bool] = Field(
        default=False,
        description=(
            "Whether the date_from and date_to should be used verbatim. Disables rounding to the start and end of"
            " period."
        ),
    )
```

[the remark this answers is not in the document]
Comment (robbie-c) [reply posthog/hogql_queries/utils/query_previous_period_date_range.py:18]: Is there any reason not to use InsightDateRange?

**Review by robbie-c — COMMENTED** [review]
Screenshot 2024-12-14 at 15 56 07

I got this error on the paths insight when I went from "no comparison" to "compare to previous period"

File: frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx (hunk 1 above)
Comment (rafaeelaudibert) [reply frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx]: I've included it :)

File: mypy-baseline.txt [hunk 5]
```diff
@@ -1,67 +1,4 @@
 posthog/warehouse/models/ssh_tunnel.py:0: error: Incompatible types in assignment (expression has type "NoEncryption", variable has type "BestAvailableEncryption")  [assignment]
-posthog/temporal/data_imports/pipelines/sql_database_v2/schema_types.py:0: error: Statement is unreachable  [unreachable]
```
[the remark this answers is not in the document]
Comment (rafaeelaudibert) [reply mypy-baseline.txt:2]: There was a single difference. I really don't know why everything was reordered 🤔

File: posthog/hogql_queries/utils/query_previous_period_date_range.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rafaeelaudibert) [reply posthog/hogql_queries/utils/query_previous_period_date_range.py:18]: Basically because of the default. Insights want to set 7 days as default, we're happy with nothing. I'll change it to have no default to improve the code readability, we can always revert it

File: posthog/hogql_queries/web_analytics/external_clicks.py (hunk 3 above)
Comment (rafaeelaudibert) [reply posthog/hogql_queries/web_analytics/external_clicks.py:53]: good catch

**Comment by rafaeelaudibert** [comment]
> I got this error on the paths insight when I went from "no comparison" to "compare to previous period"

@robbie-c I think this is happening because you have some cached data, I can't reproduce it 🤔.

I've added a failsafe to guarantee that doesn't happen, though.

File: mypy-baseline.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (robbie-c) [reply mypy-baseline.txt:2]: Yeah fair enough - don't worry too much about it. In general we should try to reduce this list but IMO we don't need to be too strict

**Review by robbie-c — APPROVED** [review]
:shipit:
